### ISSUE TITLE
openat_report_new() should not follow symlinks when creating a file

### ISSUE THREAD

**Issue by poettering** [systemd/systemd#34088]
### systemd version the issue has been seen with

git main

### CPU architectures issue was seen on

None

### Expected behaviour you didn't see

I am pretty sure the old behaviour was right here, I don't grok why this was changed.

https://github.com/systemd/systemd/pull/33825/commits/0dd82dab91eaac5e7b17bd5e9a1e07c6d2b78dca#diff-b1ee592d251974ebf48b045b61ef2c2a783dfe267da68ec6637866ec2b4181c1R1114

**Comment by DaanDeMeyer** [comment]
@poettering I think we want `openat_report_new()` to be a wrapper around `openat()` with exactly the same behavior? Otherwise you get exactly what I run into, random test failures because I switched from `openat()` to `openat_report_new()` in `xopenat()`. It's called `openat_report_new()`, not `openat_report_new_and_dont_follow_symlinks_on_ocreat()`. I think it makes sense to stick to the same behavior as `openat()` in such a helper.

**Comment by poettering** [comment]
> @poettering I think we want `openat_report_new()` to be a wrapper around `openat()` with exactly the same behavior?

well, not sure. I think it's fine to make some restrictions.

For example, maybe just add `assert(!FLAGS_SET(flags, O_CREAT) || FLAGS_SET(flags, O_NOFOLLOW))`, and add a comment: "Let's not support O_CREAT without O_NOFOLLOW, since that's a security issue and a UNIX misfeature".

Or in other words: if in our codebase we ever call openat_report_new() with O_CREAT but without O_NOFOLLOW i#d consider that a bug I am sure.

**Comment by DaanDeMeyer** [comment]
```
 775/1705 systemd:fuzz / fuzz-journal-remote_crash-source-realtime-ts       FAIL              0.15s   killed by signal 6 SIGABRT
 776/1705 systemd:fuzz / fuzz-journal-remote_oss-fuzz-21122                 FAIL              0.15s   killed by signal 6 SIGABRT
 777/1705 systemd:fuzz / fuzz-journal-remote_invalid-ts.txt                 FAIL              0.17s   killed by signal 6 SIGABRT
 778/1705 systemd:fuzz / fuzz-journal-remote_crash-nul-in-field-name        FAIL              0.18s   killed by signal 6 SIGABRT
 779/1705 systemd:fuzz / fuzz-journal-remote_oss-fuzz-25353                 FAIL              0.16s   killed by signal 6 SIGABRT
 780/1705 systemd:fuzz / fuzz-journal-remote_oss-fuzz-28817                 FAIL              0.13s   killed by signal 6 SIGABRT
 781/1705 systemd:fuzz / fuzz-journal-remote_oss-fuzz-8659                  FAIL              0.13s   killed by signal 6 SIGABRT
 782/1705 systemd:fuzz / fuzz-journal-remote_oss-fuzz-8658                  FAIL              0.16s   killed by signal 6 SIGABRT
 783/1705 systemd:fuzz / fuzz-journal-remote_oss-fuzz-9341                  FAIL              0.15s   killed by signal 6 SIGABRT
 785/1705 systemd:fuzz / fuzz-journal-remote_oss-fuzz-8686                  FAIL              0.18s   killed by signal 6 SIGABRT
 795/1705 systemd:fuzz / fuzz-journal-remote_sample.txt                     FAIL              0.17s   killed by signal 6 SIGABRT
1073/1705 systemd:libsystemd / test-journal-interleaving                    FAIL              0.12s   killed by signal 6 SIGABRT
1074/1705 systemd:libsystemd / test-journal-flush                           FAIL              0.13s   killed by signal 6 SIGABRT
1077/1705 systemd:libsystemd / test-journal-stream                          FAIL              0.15s   killed by signal 6 SIGABRT
1080/1705 systemd:libsystemd / test-journal                                 FAIL              0.17s   killed by signal 6 SIGABRT
1102/1705 systemd:test / test-copy                                          FAIL              0.12s   killed by signal 6 SIGABRT
1131/1705 systemd:test / test-fs-util                                       FAIL              0.11s   killed by signal 6 SIGABRT
1138/1705 systemd:test / test-id128                                         FAIL              0.10s   killed by signal 6 SIGABRT
```

Test failures I get from adding that assertion. I don't disagree that we need to fix this but let's not prevent using `openat_report_new()` in `xopenat()` just because we happen to use `O_CREAT` without `O_NOFOLLOW` in a bunch of code.

If you wanna go ahead and fix all the call sites to pass `O_NOFOLLOW` and then remove this bit again I'm all for it though.

**Comment by poettering** [comment]
My proposed fix waiting in #34850 now.
